### ISSUE TITLE
Difference between weighted accuracy metric of Keras and Sklearn

### ISSUE THREAD

**Issue by Oremountainflorian** [keras-team/keras#12991]
### Intro

Hej everyone, 

I am working on my diploma thesis and I face a binary classification problem with imbalanced class contribution. I have around 10 times more negative ("0") labels as positive ("1") labels. For that reason I considered not only observing accuracy and ROC-AUC, but also weighted/ balanced accuracy and Precision-Recall-AUC. 

### Issue description

During some calculations on the validation set in a custom callback I noticed, more or less by coincidence, that the weighted accuracy is always **different** from my results using _sklearn.metrics.accuracy_score()_.  

Using Keras, weighted accuracy has to be declared in _model.compile()_ and is a key in the logs{} dictionary after every epoch (and is also written to the log file by the CSVLogger callback or to the history object) or is returned as value in a list by model.evaluate(), 
```python
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'], 
              weighted_metrics=['accuracy'])
```
I calculate the _val_sample_weights_ vector based on the class contribution of the training set with the Sklearn.metrics function _class_weight.compute_sample_weight()_ and with the help of _class_weight.compute_class_weight()_.
```python
cls_weights = class_weight.compute_class_weight('balanced', np.unique(y_train._values), 
                                                y_train._values)
cls_weight_dict = {0: cls_weights[0], 1: cls_weights[1]}
val_sample_weights = class_weight.compute_sample_weight(cls_weight_dict, y_test._values)
```
In _model.fit()_ I pass this vector togehter with the validation data and to _sklearn.metrics.accuracy_score()_ I pass it to the parameter name  _sample_weight_ to compare the results on the same basis. 
```python
model_output = model.fit(x_train, y_train, epochs=500, batch_size=32, verbose=1,
                         validation_data=(x_test, y_test, val_sample_weights))
```
Furthermore I derived the equation how Sklearn computes the weighted accuracy from several easy examples and it seems that it's computed by the following equation, which seems quite reasonable to me: 

![equation](https://latex.codecogs.com/gif.latex?Acc_{weighted}=\frac{w_p&space;TP&plus;w_n&space;TN}{w_p&space;(TP&space;&plus;&space;FN)&plus;w_n&space;(TN&plus;&space;FP)}=\frac{w_p&space;TP&plus;w_n&space;TN}{TP&space;&plus;&space;FN&plus;TN&plus;&space;FP})

TP, TN, FP and FN are the values reported in the confusion matrix and w_p and w_n are the class weights of the positive and negative class respectively. 

An easy example to test it can be found here: 

[towardsdatascience.com/k-as-in-keras-simple-classification-model-a9d2d23d5b5a](url)

Just for the sake of completeness, _sklearn.metrics.accuracy_score(..., sample_weight=)_ returns the same result as _sklearn.metrics.balanced_accuracy_score()_.  

### System Information

- GeForce RTX 2080 Ti
- Keras 2.2.4
- Tensorflow-gpu 1.13.1
- Sklearn 0.19.2
- Python 3.6.8
- CUDA Version 10.0.130

### Code example

I searched an easy example to make the issue easy to reproduce, even if the class imbalance here is weaker (1:2 not 1:10). It's based on the introductory tutorial to Keras which can be found here:

[https://towardsdatascience.com/k-as-in-keras-simple-classification-model-a9d2d23d5b5a](url)

The _Pima Indianas onset diabets_ dataset will be downloaded, as done in the link above, from the repository of Jason Brownlee, the maker of the homepage Machine Learning Mastery. But I guess it can also be downloaded from various other sites. 

So finally here's the code: 

```python
from keras.layers import Dense, Dropout
from keras.models import Sequential
from keras.regularizers import l2
import pandas as pd
import numpy as np
from sklearn.utils import class_weight
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

file = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/' \
       'pima-indians-diabetes.data.csv'

# Load csv data from file to data using pandas
data = pd.read_csv(file, names=['pregnancies', 'glucose', 'diastolic', 'triceps', 'insulin',
                                'bmi', 'dpf', 'age', 'diabetes'])

# Process data
data.head()
x = data.drop(columns=['diabetes'])
y = data['diabetes']

# Split into train and test
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=0)

# define a sequential model
model = Sequential()
# 1st hidden layer
model.add(Dense(100, activation='relu', input_dim=8, kernel_regularizer=l2(0.01)))
model.add(Dropout(0.3))
# 2nd hidden layer
model.add(Dense(100, activation='relu', kernel_regularizer=l2(0.01)))
model.add(Dropout(0.3))
# Output layer
model.add(Dense(1, activation='sigmoid'))
# Compilation with weighted metrics
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'], 
                         weighted_metrics=['accuracy'])

# Calculate validation _sample_weights_ based on the class distribution of train labels and 
# apply it to test labels using Sklearn
cls_weights = class_weight.compute_class_weight('balanced', np.unique(y_train._values), 
                                                y_train._values)
cls_weight_dict = {0: cls_weights[0], 1: cls_weights[1]}
val_sample_weights = class_weight.compute_sample_weight(cls_weight_dict, y_test._values)

# Train model
model_output = model.fit(x_train, y_train, epochs=500, batch_size=32, verbose=1,
                         validation_data=(x_test, y_test, val_sample_weights))

# Predict model
y_pred = model.predict(x_test, batch_size=32, verbose=1)

# Classify predictions based on threshold at 0.5
y_pred_binary = (y_pred > 0.5) * 1

# Sklearn metrics
sklearn_accuracy = accuracy_score(y_test, y_pred_binary)
sklearn_weighted_accuracy = accuracy_score(y_test, y_pred_binary, 
                                           sample_weight=val_sample_weights)

# metric_list has 3 entries: [0] val_loss weighted by val_sample_weights, [1] val_accuracy 
# [2] val_weighted_accuracy
metric_list = model.evaluate(x_test, y_test, batch_size=32, verbose=1, 
                             sample_weight=val_sample_weights)

print('sklearn_accuracy=%.3f' %sklearn_accuracy)
print('sklearn_weighted_accuracy=%.3f' %sklearn_weighted_accuracy)
print('keras_evaluate_accuracy=%.3f' %metric_list[1])
print('keras_evaluate_weighted_accuracy=%.3f' %metric_list[2])
```

### Results and summary

For example I get:
`sklearn_accuracy=0.792`
`sklearn_weighted_accuracy=0.718`
`keras_evaluate_accuracy=0.792`
`keras_evaluate_weighted_accuracy=0.712`

The "unweighted" accuracy value is the same, both for Sklearn as for Keras. The difference isn't really big, but it grows bigger as the dataset becomes more imbalanced. For example for my task it always differs around 5% from each other! 

Maybe I'm missing something and it's supposed to be like that, but anyways it's confusing that Keras and Sklearn provide different values, especially thinking of the whole class_weights and sample_class_weights thing as a topic hard to get into.

**Comment by soi** [comment]
A similar problem came up for me. After digging in some code it appears like the function in weighted_masked_objective() seems to be indifferent to zeros. I could be totally wrong though. Would be nice to know the real reason and thanks for the thorough discription.

**Comment by Hoeze** [comment]
Are there any updates on this?

**Comment by Ram-WD** [comment]
any update ?
